### PR TITLE
[Model Runner V2] Minor refactoring for penalties

### PR DESCRIPTION
**Pull request by WoosukKwon**
Now that we have `prompt_lens` as a GPU tensor, `bincount` can cover multiple requests with a single kernel launch.
Plus, small refactoring to align the style with `BadWordsState`.

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/penalties.py [hunk 1]
```diff
@@ -214,51 +228,82 @@ def apply_penalties(
...
+    prompt_len: torch.Tensor,
+    prefill_len: torch.Tensor,
     prompt_bin_mask: torch.Tensor,
     output_bin_counts: torch.Tensor,
+    max_model_len: int,
 ) -> None:
-    prompt_bin_mask.zero_()
-    output_bin_counts.zero_()
+    prompt_bin_mask[idx_mapping].zero_()
+    output_bin_counts[idx_mapping].zero_()
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/penalties.py]: Indexing a tensor like this will create a new tensor unless it's an assignment (so this won't actually modify the original tensors)

```suggestion
    prompt_bin_mask[idx_mapping] = 0
    output_bin_counts[idx_mapping] = 0
```

File: vllm/v1/worker/gpu/sample/penalties.py [hunk 2]
```diff
@@ -49,22 +61,24 @@ def add_request(self, req_idx: int, sampling_params: SamplingParams) -> None:
...
-                int(prompt_lens[req_idx]),
-                self.prompt_bin_mask[req_idx],
-                self.output_bin_counts[req_idx],
+                idx_mapping,
+                self.all_token_ids,
+                self.prompt_len,
+                self.prefill_len,
+                self.prompt_bin_mask,
+                self.output_bin_counts,
+                self.max_model_len,
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/penalties.py]: Instead of `max_model_len` could instead here pass `int(prefill_len.np[self._penalties_reqs].max())`.

In general that should be much smaller I think?

(need to share `prefill_len.np` with sampler for this of course)

**Review by njhill — COMMENTED** [review]
Nice changes!

One bug and one suggestion

File: vllm/v1/worker/gpu/sample/penalties.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/sample/penalties.py]: OMG. Thanks for catching this.

File: vllm/v1/worker/gpu/sample/penalties.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/sample/penalties.py]: Actually, it was totally wrong to only pass the GPU tensor, because `prefill_len` and `prompt_len` use double buffering. We should pass the entire `UvaBackedTensor`...

Comment (njhill) [reply vllm/v1/worker/gpu/sample/penalties.py]: Oops ... yes this is true for the three tensors added to sampler constructor in https://github.com/vllm-project/vllm/pull/33433 I guess. Maybe best for `BadWordsState` and `PenaltiesState` to not keep this state and just pass the tensors to e.g. `apply_bad_words()`?

Comment (WoosukKwon) [reply vllm/v1/worker/gpu/sample/penalties.py]: @njhill 🤔 For now, I’m passing `RequestState` into the constructors of `Sampler`, `PenaltiesState`, and `BadWordsState`. My thinking was that the generic logits processor API we’re considering would likely need access to these states anyway.

Comment (njhill) [reply vllm/v1/worker/gpu/sample/penalties.py]: SGTM we can easily refactor if/as needed

**Review by njhill — APPROVED** [review]

File: vllm/v1/worker/gpu/sample/penalties.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/sample/penalties.py]: Yeah happy to revisit. The code here is pretty much error-prone.
